### ISSUE TITLE
Null character in buffer is sent as newline

### ISSUE THREAD

**Issue by ii41** [prabirshrestha/vim-lsp#1435]
`lsp#utils#buffer#_get_lines` uses `getbufline` to get buffer contents:

https://github.com/prabirshrestha/vim-lsp/blob/296fb98d198cbbb5c5c937c09b84c8c7a9605a16/autoload/lsp/utils/buffer.vim#L34

As neovim/vim inserts null characters for multiline completion (https://github.com/vim/vim/issues/2505 https://github.com/neovim/neovim/issues/7769), null characters keep appearing in the buffers today with language servers sending multiline completion items. `getbufline` converts null characters in the buffer into newlines. The behavior is planned to be not changed, as changing this will break the vimscript type system, which doesn't allow null bytes in strings. See https://github.com/neovim/neovim/issues/22093.

As vim-lsp does not do special processing for such a special case before sending the text changes to the server, a newline is sent when there is actually a null. This results in language servers not being able to work correctly (and has caused me to wrongly report to pyright: https://github.com/microsoft/pyright/issues/4559, as pyright provides multiline completion for `__init__`).

**Comment by mattn** [comment]
Thanks. Could you please check below's small patch fixies this?
```
let l:lines = map(getbufline(a:buf, 1, '$'), 'substitute(v:val, "\n", " ", "g")')
```

**Comment by ii41** [comment]
I tried it out. Yeah, it kind of fixes this... It's apparently not a perfect solution: lsp won't correctly report that there's an invalid character, but at least when the null is removed lsp works correctly.

This problem doesn't hurt my daily usage even without any patching as I use an `asyncomplete.vim` preprocessor to make multiline completion items into one-line ones (the follow-up lines are not really useful anyway), and normally of course do not interact with code containing null bytes.

I have also been thinking about how this can be completely fixed since I reported this, and found that it was really hard, and possibly not possible in pure vimscript and when both vim and neovim need to be supported... From a practical perspective, yeah, a patch like this is fine. I think perhaps
```
let l:lines = map(getbufline(a:buf, 1, '$'), 'substitute(v:val, "\n", "\ufffd", "g")')
```
is a better solution, so that when something goes wrong at least the user knows that it's wrong. Leaving a message could be even better.
